### PR TITLE
Paged Attention KVCacheManager

### PR DESCRIPTION
**Pull request by JRosenkranz**
This PR is part of a stack of PRs related to paged attention + speculative decoding:

- [ ] Paged Attention KVCache
- [ ] Paged Model
- [ ] Speculative generation

Full implementation of the above can be found here: https://github.com/foundation-model-stack/fms-extras/pull/7

In this PR, a KVCacheManager for Paged Attention is introduced which takes advantage of the paged attention cuda kernels. The manager does logical allocation of tokens as well as garbage collection.

### CODE REVIEW

File: fms_extras/utils/cache/__init__.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    num_tokens_per_sequence: List[int], context_lengths: Optional[List[int]] = None
+) -> List[List[int]]:
+    """Compute position ids based on the current context lengths and the new tokens to add
+
+    Parameters
+    ----------
+    num_tokens_per_sequence: List[int]
+        number of tokens to be added to each sequence
+    context_lengths: List[int], optional
+        optional list of current context lengths per sequence. If none, will assume no context length and starting
```
Comment (nairbv) [inline fms_extras/utils/cache/__init__.py]: the explanation here seems to describe the length of the context (prompt?) being provided to the model... but the code below seems to suggest that this is the start_idx of each row after padding. I think there are a lot of ways to interpret this so we should try to make it less ambiguous.

File: fms_extras/utils/cache/__init__.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
+import abc
+import dataclasses
+from typing import List, Optional, Tuple
+
+import torch
+
+
+def compute_position_ids(
```
Comment (nairbv) [inline fms_extras/utils/cache/__init__.py]: we should probably have a unit test for this function?

File: fms_extras/utils/cache/__init__.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        cache data is in pre-fill
+    """
+
+    data: List[Tuple[torch.Tensor, torch.Tensor]]
+    max_sequence_length: int
+    sequence_ids: List[int]
+    context_lengths: Optional[torch.Tensor]
+
+    def compute_position_ids(self, num_tokens_per_sequence: List[int]) -> torch.Tensor:
+        position_ids_list = compute_position_ids(
```
Comment (nairbv) [inline fms_extras/utils/cache/__init__.py]: this is a bit confusing, looks as if it's recursive but isn't

File: fms_extras/utils/cache/__init__.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    position_ids = []
+    for seq_i, num_tokens in enumerate(num_tokens_per_sequence):
+        if context_lengths is None:
+            start = 0
+        else:
+            start = context_lengths[seq_i] - 1
+        position_ids_i = [0 for _ in range(max_tokens - num_tokens)] + [
+            i for i in range(start, start + num_tokens)
+        ]
+        position_ids.append(position_ids_i)
```
Comment (nairbv) [inline fms_extras/utils/cache/__init__.py]: would this all be simpler as tensor operations instead of for loops?

File: fms_extras/utils/cache/__init__.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    return position_ids
+
+
+class AttentionComputationMixin(metaclass=abc.ABCMeta):
```
Comment (nairbv) [inline fms_extras/utils/cache/__init__.py]: seems this type isn't really used? one class mixes it in, but I'm not seeing polymorphism, and the actual switching between attention mechanisms is done with conditional statements.

File: fms_extras/utils/cache/__init__.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+                r - l + 1
+                for l, r in zip(num_tokens_per_sequence, self.context_lengths.tolist())
+            ],
+        )
+        return torch.tensor(
+            position_ids_list, dtype=torch.long, device=self.data[0][0].device
+        )
+
+
+class KVCacheManager(metaclass=abc.ABCMeta):
```
Comment (nairbv) [inline fms_extras/utils/cache/__init__.py]: likewise here -- I think we were initially considering building abstractions where we'd have regular kv_cache's managed by a class like this, and might be able to swap that out for alternatives. If we only have one subclass though, this parent class doesn't seem necessary.

File: csrc/paged_attention/attention/attention_generic.cuh [hunk 7]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Adapted from https://github.com/NVIDIA/FasterTransformer/blob/release/v5.3_tag/src/fastertransformer/kernels/decoder_masked_multihead_attention_utils.h
```
Comment (nairbv) [inline csrc/paged_attention/attention/attention_generic.cuh:2]: Have we changed anything in any of these files or is everything under paged_attention copy/paste from vllm?
If it's all copy/paste, can we use some kind of git submodule instead?
If not, can we highlight where exactly we got it from and at what commit hash? Maybe we could add a top level README in the paged_attention directory?

If we have changed anything, we should highlight what has changed, and possibly add unit tests for those changes. If we expect to change anything in the future, we may need to import any existing tests of this code so that we can ensure we don't break it.

File: fms_extras/utils/cache/__init__.py (hunk 5 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/__init__.py]: This was left over from when we were going to implement a version of flash computation, or if another form comes along and we want to use the same MHA. Because I kept this PR to only `paged`, we probably don't need this for this PR. Having said that, if we ever use these structures for other forms of attention in the future, it might make sense to keep it

File: fms_extras/utils/cache/__init__.py (hunk 6 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/__init__.py]: Same answer as above, I can remove for this PR, but figured it was best to keep it in case we want to create a Flash KVCacheManager or some other form

File: csrc/paged_attention/attention/attention_generic.cuh (hunk 7 above)
Comment (JRosenkranz) [reply csrc/paged_attention/attention/attention_generic.cuh:2]: This was copy and pasted from vLLM, however we have removed a lot of it that we are not using (I'll find the exact version this was from) I think we keep it here as is, and put a comment for what version of vllm it is. One reason for this is I anticipate we will be making some changes in the future.

File: fms_extras/utils/cache/__init__.py (hunk 3 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/__init__.py]: Removed the extra utility function as it is not used.

File: fms_extras/utils/cache/__init__.py (hunk 2 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/__init__.py]: This is being tested through test_paged now (removed the utility function as it wasn't used)

File: fms_extras/utils/cache/__init__.py (hunk 5 above)
Comment (nairbv) [reply fms_extras/utils/cache/__init__.py]: I think we should probably remove until/unless we have a use-case for it. it would be easy to re-add when/if we decide to build something that way

File: fms_extras/utils/cache/paged.py [hunk 8]
```diff
@@ -0,0 +1,1044 @@
...
+from fms_extras.utils.cache import (
+    AttentionComputationMixin,
+    CacheDataLayer,
+    CacheDataWithMetadata,
+    KVCache,
+    KVCacheManager,
+)
+
+
+lib = torch.library.Library("paged_attention", "FRAGMENT")
```
Comment (nairbv) [inline fms_extras/utils/cache/paged.py:18]: what's the purpose of putting our operators in the torch namespace? I'm assuming it's something to do with making this compile, but we might want a comment explaining the motive

File: fms_extras/utils/cache/paged.py [hunk 9]
```diff
@@ -0,0 +1,1044 @@
...
+    return out
+
+
+lowering.fallbacks.add(torch.ops.paged_attention.paged_attention_v1)
+
+
+@lowering.register_lowering(
+    torch.ops.paged_attention.paged_attention_v1, type_promotion_kind=None
+)
+def _paged_attention_v1_lowering(
```
Comment (nairbv) [inline fms_extras/utils/cache/paged.py:218]: what's the difference between v1 and v2 paged attention?

File: fms_extras/utils/cache/__init__.py (hunk 1 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/__init__.py]: This has been removed in latest commit

File: fms_extras/utils/cache/__init__.py (hunk 5 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/__init__.py]: Ok will do

File: fms_extras/utils/cache/paged.py (hunk 8 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/paged.py:18]: yes, that is the reason, ill add a comment

File: fms_extras/utils/cache/paged.py (hunk 9 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/paged.py:218]: We are doing same as vLLM for this call. Based on the comments in vLLM:

```
# NOTE(woosuk): We use a simple heuristic to decide whether to use
# PagedAttention V1 or V2. If the number of partitions is 1, we use
# V1 to avoid the overhead of reduction. Also, if the number of
# sequences or heads is large, we use V1 since there is enough work
# to parallelize.
# TODO(woosuk): Tune this heuristic.
# For context len > 8192, use V2 kernel to avoid shared memory shortage.
```

File: csrc/paged_attention/attention/attention_generic.cuh (hunk 7 above)
Comment (JRosenkranz) [reply csrc/paged_attention/attention/attention_generic.cuh:2]: This was from version 0.2.7, I added a comment in the kernels.

File: fms_extras/utils/cache/__init__.py (hunk 6 above)
Comment (JRosenkranz) [reply fms_extras/utils/cache/__init__.py]: removed for now til we have another implementation

File: csrc/paged_attention/attention/attention_generic.cuh (hunk 7 above)
Comment (nairbv) [reply csrc/paged_attention/attention/attention_generic.cuh:2]: if I control-f I'm not seeing "0.2.7" anywhere. maybe a short README in the `paged_attention` directory would help explain where it came from / how to maintain?

File: requirements-build.txt [hunk 10]
```diff
@@ -0,0 +1,6 @@
+# Should be mirrored in pyproject.toml
+ninja
+packaging
+setuptools>=49.4.0
+torch==2.2.0
```
Comment (nairbv) [inline requirements-build.txt:5]: duplicate requirement? I assume whenever we install requirements-build we also first install requirements.

File: requirements.txt [hunk 11]
```diff
@@ -0,0 +1,7 @@
+# Used to install pinned dependencies
+# Useful for dev/test jobs caches
+# Must be kept in sync with setup.py
+torch == 2.2.0 # This is what is installed in CI today
```
Comment (nairbv) [inline requirements.txt:4]: I was surprised we didn't do >= but I see we have == in fms too.

**Review by nairbv — APPROVED** [review]
looks fine, mostly copy/paste code from elsewhere so didn't read every line closely.

I'd suggest putting some more detail somewhere about how to maintain the vLLM kernels before landing though. We could add a README pointing out the specific version/commit/tag the code came from, and propose some common way to refer to any changes we make to the copy/pasted code.
E.g. we might propose something like comments including `//<fms>` and `//</fms>` around changes, and could add a place to log/explain our adaptations in that README.
